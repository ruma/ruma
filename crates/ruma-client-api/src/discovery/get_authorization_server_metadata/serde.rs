use std::collections::BTreeSet;

use serde::{de, Deserialize};
use url::Url;

use super::msc2965::{
    AccountManagementAction, AuthorizationServerMetadata, CodeChallengeMethod, GrantType, Prompt,
    ResponseMode, ResponseType,
};

impl<'de> Deserialize<'de> for AuthorizationServerMetadata {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        let helper = AuthorizationServerMetadataDeHelper::deserialize(deserializer)?;

        let AuthorizationServerMetadataDeHelper {
            issuer,
            authorization_endpoint,
            token_endpoint,
            registration_endpoint,
            response_types_supported,
            response_modes_supported,
            grant_types_supported,
            revocation_endpoint,
            code_challenge_methods_supported,
            account_management_uri,
            account_management_actions_supported,
            device_authorization_endpoint,
            prompt_values_supported,
        } = helper;

        // Require `code` in `response_types_supported`.
        if !response_types_supported.contains(&ResponseType::Code) {
            return Err(de::Error::custom("missing value `code` in `response_types_supported`"));
        }

        // Require `query` and `fragment` in `response_modes_supported`.
        if let Some(response_modes) = &response_modes_supported {
            let query_found = response_modes.contains(&ResponseMode::Query);
            let fragment_found = response_modes.contains(&ResponseMode::Fragment);

            if !query_found && !fragment_found {
                return Err(de::Error::custom(
                    "missing values `query` and `fragment` in `response_modes_supported`",
                ));
            }
            if !query_found {
                return Err(de::Error::custom(
                    "missing value `query` in `response_modes_supported`",
                ));
            }
            if !fragment_found {
                return Err(de::Error::custom(
                    "missing value `fragment` in `response_modes_supported`",
                ));
            }
        }
        // If the field is missing, the default value is `["query", "fragment"]`, according to
        // RFC8414.
        let response_modes_supported = response_modes_supported
            .unwrap_or_else(|| [ResponseMode::Query, ResponseMode::Fragment].into());

        // Require `authorization_code` and `refresh_token` in `grant_types_supported`.
        let authorization_code_found =
            grant_types_supported.contains(&GrantType::AuthorizationCode);
        let refresh_token_found = grant_types_supported.contains(&GrantType::RefreshToken);
        if !authorization_code_found && !refresh_token_found {
            return Err(de::Error::custom(
                    "missing values `authorization_code` and `refresh_token` in `grant_types_supported`",
                ));
        }
        if !authorization_code_found {
            return Err(de::Error::custom(
                "missing value `authorization_code` in `grant_types_supported`",
            ));
        }
        if !refresh_token_found {
            return Err(de::Error::custom(
                "missing value `refresh_token` in `grant_types_supported`",
            ));
        }

        // Require `S256` in `code_challenge_methods_supported`.
        if !code_challenge_methods_supported.contains(&CodeChallengeMethod::S256) {
            return Err(de::Error::custom(
                "missing value `s256` in `code_challenge_methods_supported`",
            ));
        }

        Ok(AuthorizationServerMetadata {
            issuer,
            authorization_endpoint,
            token_endpoint,
            registration_endpoint,
            response_types_supported,
            response_modes_supported,
            grant_types_supported,
            revocation_endpoint,
            code_challenge_methods_supported,
            account_management_uri,
            account_management_actions_supported,
            device_authorization_endpoint,
            prompt_values_supported,
        })
    }
}

#[derive(Deserialize)]
struct AuthorizationServerMetadataDeHelper {
    issuer: Url,
    authorization_endpoint: Url,
    token_endpoint: Url,
    registration_endpoint: Option<Url>,
    response_types_supported: BTreeSet<ResponseType>,
    response_modes_supported: Option<BTreeSet<ResponseMode>>,
    grant_types_supported: BTreeSet<GrantType>,
    revocation_endpoint: Url,
    code_challenge_methods_supported: BTreeSet<CodeChallengeMethod>,
    account_management_uri: Option<Url>,
    #[serde(default)]
    account_management_actions_supported: BTreeSet<AccountManagementAction>,
    device_authorization_endpoint: Option<Url>,
    #[serde(default)]
    prompt_values_supported: Vec<Prompt>,
}

#[cfg(test)]
mod tests {
    use as_variant::as_variant;
    use serde_json::{from_value as from_json_value, value::Map as JsonMap, Value as JsonValue};
    use url::Url;

    use crate::discovery::get_authorization_server_metadata::{
        msc2965::{
            AccountManagementAction, AuthorizationServerMetadata, CodeChallengeMethod, GrantType,
            ResponseMode, ResponseType,
        },
        tests::authorization_server_metadata_json,
    };

    /// A valid `AuthorizationServerMetadata` with all fields and values, as a JSON object.
    fn authorization_server_metadata_object() -> JsonMap<String, JsonValue> {
        as_variant!(authorization_server_metadata_json(), JsonValue::Object).unwrap()
    }

    /// Get a mutable reference to the array value with the given key in the given object.
    ///
    /// Panics if the property doesn't exist or is not an array.
    fn get_mut_array<'a>(
        object: &'a mut JsonMap<String, JsonValue>,
        key: &str,
    ) -> &'a mut Vec<JsonValue> {
        object.get_mut(key).unwrap().as_array_mut().unwrap()
    }

    #[test]
    fn metadata_all_fields() {
        let metadata_object = authorization_server_metadata_object();
        let metadata =
            from_json_value::<AuthorizationServerMetadata>(metadata_object.into()).unwrap();

        assert_eq!(metadata.issuer.as_str(), "https://server.local/");
        assert_eq!(metadata.authorization_endpoint.as_str(), "https://server.local/authorize");
        assert_eq!(metadata.token_endpoint.as_str(), "https://server.local/token");
        assert_eq!(
            metadata.registration_endpoint.as_ref().map(Url::as_str),
            Some("https://server.local/register")
        );

        assert_eq!(metadata.response_types_supported.len(), 1);
        assert!(metadata.response_types_supported.contains(&ResponseType::Code));

        assert_eq!(metadata.response_modes_supported.len(), 2);
        assert!(metadata.response_modes_supported.contains(&ResponseMode::Query));
        assert!(metadata.response_modes_supported.contains(&ResponseMode::Fragment));

        assert_eq!(metadata.grant_types_supported.len(), 2);
        assert!(metadata.grant_types_supported.contains(&GrantType::AuthorizationCode));
        assert!(metadata.grant_types_supported.contains(&GrantType::RefreshToken));

        assert_eq!(metadata.revocation_endpoint.as_str(), "https://server.local/revoke");

        assert_eq!(metadata.code_challenge_methods_supported.len(), 1);
        assert!(metadata.code_challenge_methods_supported.contains(&CodeChallengeMethod::S256));

        assert_eq!(
            metadata.account_management_uri.as_ref().map(Url::as_str),
            Some("https://server.local/account")
        );
        assert_eq!(metadata.account_management_actions_supported.len(), 6);
        assert!(metadata
            .account_management_actions_supported
            .contains(&AccountManagementAction::Profile));
        assert!(metadata
            .account_management_actions_supported
            .contains(&AccountManagementAction::SessionsList));
        assert!(metadata
            .account_management_actions_supported
            .contains(&AccountManagementAction::SessionView));
        assert!(metadata
            .account_management_actions_supported
            .contains(&AccountManagementAction::SessionEnd));
        assert!(metadata
            .account_management_actions_supported
            .contains(&AccountManagementAction::AccountDeactivate));
        assert!(metadata
            .account_management_actions_supported
            .contains(&AccountManagementAction::CrossSigningReset));

        assert_eq!(
            metadata.device_authorization_endpoint.as_ref().map(Url::as_str),
            Some("https://server.local/device")
        );
    }

    #[test]
    fn metadata_no_optional_fields() {
        let mut metadata_object = authorization_server_metadata_object();
        assert!(metadata_object.remove("registration_endpoint").is_some());
        assert!(metadata_object.remove("response_modes_supported").is_some());
        assert!(metadata_object.remove("account_management_uri").is_some());
        assert!(metadata_object.remove("account_management_actions_supported").is_some());
        assert!(metadata_object.remove("device_authorization_endpoint").is_some());

        let metadata =
            from_json_value::<AuthorizationServerMetadata>(metadata_object.into()).unwrap();

        assert_eq!(metadata.issuer.as_str(), "https://server.local/");
        assert_eq!(metadata.authorization_endpoint.as_str(), "https://server.local/authorize");
        assert_eq!(metadata.token_endpoint.as_str(), "https://server.local/token");
        assert_eq!(metadata.registration_endpoint, None);

        assert_eq!(metadata.response_types_supported.len(), 1);
        assert!(metadata.response_types_supported.contains(&ResponseType::Code));

        assert_eq!(metadata.response_modes_supported.len(), 2);
        assert!(metadata.response_modes_supported.contains(&ResponseMode::Query));
        assert!(metadata.response_modes_supported.contains(&ResponseMode::Fragment));

        assert_eq!(metadata.grant_types_supported.len(), 2);
        assert!(metadata.grant_types_supported.contains(&GrantType::AuthorizationCode));
        assert!(metadata.grant_types_supported.contains(&GrantType::RefreshToken));

        assert_eq!(metadata.revocation_endpoint.as_str(), "https://server.local/revoke");

        assert_eq!(metadata.code_challenge_methods_supported.len(), 1);
        assert!(metadata.code_challenge_methods_supported.contains(&CodeChallengeMethod::S256));

        assert_eq!(metadata.account_management_uri, None);
        assert_eq!(metadata.account_management_actions_supported.len(), 0);

        assert_eq!(metadata.device_authorization_endpoint, None);
    }

    #[test]
    fn metadata_additional_values() {
        let mut metadata_object = authorization_server_metadata_object();
        get_mut_array(&mut metadata_object, "response_types_supported").push("custom".into());
        get_mut_array(&mut metadata_object, "response_modes_supported").push("custom".into());
        get_mut_array(&mut metadata_object, "grant_types_supported").push("custom".into());
        get_mut_array(&mut metadata_object, "code_challenge_methods_supported")
            .push("custom".into());
        get_mut_array(&mut metadata_object, "account_management_actions_supported")
            .push("custom".into());

        let metadata =
            from_json_value::<AuthorizationServerMetadata>(metadata_object.into()).unwrap();

        assert_eq!(metadata.issuer.as_str(), "https://server.local/");
        assert_eq!(metadata.authorization_endpoint.as_str(), "https://server.local/authorize");
        assert_eq!(metadata.token_endpoint.as_str(), "https://server.local/token");
        assert_eq!(
            metadata.registration_endpoint.as_ref().map(Url::as_str),
            Some("https://server.local/register")
        );

        assert_eq!(metadata.response_types_supported.len(), 2);
        assert!(metadata.response_types_supported.contains(&ResponseType::Code));
        assert!(metadata.response_types_supported.contains(&ResponseType::from("custom")));

        assert_eq!(metadata.response_modes_supported.len(), 3);
        assert!(metadata.response_modes_supported.contains(&ResponseMode::Query));
        assert!(metadata.response_modes_supported.contains(&ResponseMode::Fragment));
        assert!(metadata.response_modes_supported.contains(&ResponseMode::from("custom")));

        assert_eq!(metadata.grant_types_supported.len(), 3);
        assert!(metadata.grant_types_supported.contains(&GrantType::AuthorizationCode));
        assert!(metadata.grant_types_supported.contains(&GrantType::RefreshToken));
        assert!(metadata.grant_types_supported.contains(&GrantType::from("custom")));

        assert_eq!(metadata.revocation_endpoint.as_str(), "https://server.local/revoke");

        assert_eq!(metadata.code_challenge_methods_supported.len(), 2);
        assert!(metadata.code_challenge_methods_supported.contains(&CodeChallengeMethod::S256));
        assert!(metadata
            .code_challenge_methods_supported
            .contains(&CodeChallengeMethod::from("custom")));

        assert_eq!(
            metadata.account_management_uri.as_ref().map(Url::as_str),
            Some("https://server.local/account")
        );
        assert_eq!(metadata.account_management_actions_supported.len(), 7);
        assert!(metadata
            .account_management_actions_supported
            .contains(&AccountManagementAction::Profile));
        assert!(metadata
            .account_management_actions_supported
            .contains(&AccountManagementAction::SessionsList));
        assert!(metadata
            .account_management_actions_supported
            .contains(&AccountManagementAction::SessionView));
        assert!(metadata
            .account_management_actions_supported
            .contains(&AccountManagementAction::SessionEnd));
        assert!(metadata
            .account_management_actions_supported
            .contains(&AccountManagementAction::AccountDeactivate));
        assert!(metadata
            .account_management_actions_supported
            .contains(&AccountManagementAction::CrossSigningReset));
        assert!(metadata
            .account_management_actions_supported
            .contains(&AccountManagementAction::from("custom")));

        assert_eq!(
            metadata.device_authorization_endpoint.as_ref().map(Url::as_str),
            Some("https://server.local/device")
        );
    }

    #[test]
    fn metadata_missing_required_fields() {
        let original_metadata_object = authorization_server_metadata_object();

        let mut metadata_object = original_metadata_object.clone();
        assert!(metadata_object.remove("issuer").is_some());
        from_json_value::<AuthorizationServerMetadata>(metadata_object.into()).unwrap_err();

        let mut metadata_object = original_metadata_object.clone();
        assert!(metadata_object.remove("authorization_endpoint").is_some());
        from_json_value::<AuthorizationServerMetadata>(metadata_object.into()).unwrap_err();

        let mut metadata_object = original_metadata_object.clone();
        assert!(metadata_object.remove("token_endpoint").is_some());
        from_json_value::<AuthorizationServerMetadata>(metadata_object.into()).unwrap_err();

        let mut metadata_object = original_metadata_object.clone();
        assert!(metadata_object.remove("response_types_supported").is_some());
        from_json_value::<AuthorizationServerMetadata>(metadata_object.into()).unwrap_err();

        let mut metadata_object = original_metadata_object.clone();
        assert!(metadata_object.remove("grant_types_supported").is_some());
        from_json_value::<AuthorizationServerMetadata>(metadata_object.into()).unwrap_err();

        let mut metadata_object = original_metadata_object.clone();
        assert!(metadata_object.remove("revocation_endpoint").is_some());
        from_json_value::<AuthorizationServerMetadata>(metadata_object.into()).unwrap_err();

        let mut metadata_object = original_metadata_object;
        assert!(metadata_object.remove("code_challenge_methods_supported").is_some());
        from_json_value::<AuthorizationServerMetadata>(metadata_object.into()).unwrap_err();
    }

    #[test]
    fn metadata_missing_required_values() {
        let original_metadata_object = authorization_server_metadata_object();

        let mut metadata_object = original_metadata_object.clone();
        get_mut_array(&mut metadata_object, "response_types_supported").clear();
        from_json_value::<AuthorizationServerMetadata>(metadata_object.into()).unwrap_err();

        let mut metadata_object = original_metadata_object.clone();
        get_mut_array(&mut metadata_object, "response_modes_supported").clear();
        from_json_value::<AuthorizationServerMetadata>(metadata_object.into()).unwrap_err();

        let mut metadata_object = original_metadata_object.clone();
        get_mut_array(&mut metadata_object, "response_modes_supported").remove(0);
        from_json_value::<AuthorizationServerMetadata>(metadata_object.into()).unwrap_err();

        let mut metadata_object = original_metadata_object.clone();
        get_mut_array(&mut metadata_object, "response_modes_supported").remove(1);
        from_json_value::<AuthorizationServerMetadata>(metadata_object.into()).unwrap_err();

        let mut metadata_object = original_metadata_object.clone();
        get_mut_array(&mut metadata_object, "grant_types_supported").clear();
        from_json_value::<AuthorizationServerMetadata>(metadata_object.into()).unwrap_err();

        let mut metadata_object = original_metadata_object.clone();
        get_mut_array(&mut metadata_object, "grant_types_supported").remove(0);
        from_json_value::<AuthorizationServerMetadata>(metadata_object.into()).unwrap_err();

        let mut metadata_object = original_metadata_object.clone();
        get_mut_array(&mut metadata_object, "grant_types_supported").remove(1);
        from_json_value::<AuthorizationServerMetadata>(metadata_object.into()).unwrap_err();

        let mut metadata_object = original_metadata_object;
        get_mut_array(&mut metadata_object, "code_challenge_methods_supported").clear();
        from_json_value::<AuthorizationServerMetadata>(metadata_object.into()).unwrap_err();
    }
}
