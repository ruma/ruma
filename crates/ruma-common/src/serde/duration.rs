//! De-/serialization functions for `std::time::Duration` objects

pub mod opt_ms;
pub mod secs;
