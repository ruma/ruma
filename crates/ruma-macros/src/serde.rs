//! Methods and types for (de)serialization.

pub mod attr;
pub mod case;
pub mod deserialize_from_cow_str;
pub mod display_as_ref_str;
pub mod enum_as_ref_str;
pub mod enum_from_string;
pub mod eq_as_ref_str;
pub mod incoming;
pub mod ord_as_ref_str;
pub mod serialize_as_ref_str;
mod util;
