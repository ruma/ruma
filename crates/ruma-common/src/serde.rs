//! (De)serialization helpers for other Ruma crates.
//!
//! Part of that is a fork of [serde_urlencoded], with support for sequences in `Deserialize` /
//! `Serialize` structs (e.g. `Vec<Something>`) that are (de)serialized as `field=val1&field=val2`.
//!
//! [serde_urlencoded]: https://github.com/nox/serde_urlencoded

use serde::{de, Deserialize};
use serde_json::{value::RawValue as RawJsonValue, Value as JsonValue};

pub mod base64;
mod buf;
pub mod can_be_empty;
mod canonical_json;
mod cow;
pub mod duration;
mod empty;
pub mod json_string;
mod raw;
pub mod single_element_seq;
mod strings;
pub mod test;
pub mod urlencoded;

pub use self::{
    base64::{Base64, Base64DecodeError},
    buf::{json_to_buf, slice_to_buf},
    can_be_empty::{is_empty, CanBeEmpty},
    canonical_json::{
        to_canonical_value, try_from_json_map,
        value::{CanonicalJsonValue, Object as CanonicalJsonObject},
        Error as CanonicalJsonError,
    },
    cow::deserialize_cow_str,
    empty::vec_as_map_of_empty,
    raw::Raw,
    strings::{
        btreemap_deserialize_v1_powerlevel_values, deserialize_v1_powerlevel, empty_string_as_none,
        none_as_empty_string,
    },
};

/// The inner type of [`JsonValue::Object`].
pub type JsonObject = serde_json::Map<String, JsonValue>;

/// Check whether a value is equal to its default value.
pub fn is_default<T: Default + PartialEq>(val: &T) -> bool {
    *val == T::default()
}

/// Simply returns `true`.
///
/// Useful for `#[serde(default = ...)]`.
pub fn default_true() -> bool {
    true
}

/// Simply dereferences the given bool.
///
/// Useful for `#[serde(skip_serializing_if = ...)]`.
#[allow(clippy::trivially_copy_pass_by_ref)]
pub fn is_true(b: &bool) -> bool {
    *b
}

/// Helper function for `serde_json::value::RawValue` deserialization.
pub fn from_raw_json_value<'a, T, E>(val: &'a RawJsonValue) -> Result<T, E>
where
    T: Deserialize<'a>,
    E: de::Error,
{
    serde_json::from_str(val.get()).map_err(E::custom)
}

pub use ruma_macros::{
    AsRefStr, DeserializeFromCowStr, DisplayAsRefStr, FromString, Incoming, OrdAsRefStr,
    PartialEqAsRefStr, PartialOrdAsRefStr, SerializeAsRefStr, StringEnum, _FakeDeriveSerde,
};
