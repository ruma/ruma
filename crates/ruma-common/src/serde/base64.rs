//!Transparent base64 encoding / decoding as part of (de)serialization.

use std::{fmt, marker::PhantomData};

use serde::{de, Deserialize, Deserializer, Serialize, Serializer};

/// A wrapper around `B` (usually `Vec<u8>`) that (de)serializes from / to a base64 string.
///
/// The base64 character set (and miscellaneous other encoding / decoding options) can be customized
/// through the generic parameter `C`.
#[derive(Clone, PartialEq, Eq, PartialOrd, Ord)]
pub struct Base64<C = Standard, B = Vec<u8>> {
    bytes: B,
    // Invariant PhantomData, Send + Sync
    _phantom_conf: PhantomData<fn(C) -> C>,
}

/// Config used for the [`Base64`] type.
pub trait Base64Config {
    /// The config as a constant.
    ///
    /// Opaque so our interface is not tied to the base64 crate version.
    #[doc(hidden)]
    const CONF: Conf;
}

#[doc(hidden)]
pub struct Conf(base64::Config);

/// Standard base64 character set without padding.
///
/// Allows trailing bits in decoding for maximum compatibility.
#[non_exhaustive]
// Easier than implementing these all for Base64 manually to avoid the `C: Trait` bounds.
#[derive(Clone, PartialEq, Eq, PartialOrd, Ord)]
pub struct Standard;

impl Base64Config for Standard {
    // See https://github.com/matrix-org/matrix-spec/issues/838
    const CONF: Conf = Conf(base64::STANDARD_NO_PAD.decode_allow_trailing_bits(true));
}

/// Url-safe base64 character set without padding.
///
/// Allows trailing bits in decoding for maximum compatibility.
#[non_exhaustive]
// Easier than implementing these all for Base64 manually to avoid the `C: Trait` bounds.
#[derive(Clone, PartialEq, Eq, PartialOrd, Ord)]
pub struct UrlSafe;

impl Base64Config for UrlSafe {
    const CONF: Conf = Conf(base64::URL_SAFE_NO_PAD.decode_allow_trailing_bits(true));
}

impl<C: Base64Config, B: AsRef<[u8]>> Base64<C, B> {
    /// Create a `Base64` instance from raw bytes, to be base64-encoded in serialialization.
    pub fn new(bytes: B) -> Self {
        Self { bytes, _phantom_conf: PhantomData }
    }

    /// Get a reference to the raw bytes held by this `Base64` instance.
    pub fn as_bytes(&self) -> &[u8] {
        self.bytes.as_ref()
    }

    /// Encode the bytes contained in this `Base64` instance to unpadded base64.
    pub fn encode(&self) -> String {
        base64::encode_config(&self.bytes, C::CONF.0)
    }
}

impl<C, B> Base64<C, B> {
    /// Get the raw bytes held by this `Base64` instance.
    pub fn into_inner(self) -> B {
        self.bytes
    }
}

impl<C: Base64Config> Base64<C> {
    /// Create a `Base64` instance containing an empty `Vec<u8>`.
    pub fn empty() -> Self {
        Self::new(Vec::new())
    }

    /// Parse some base64-encoded data to create a `Base64` instance.
    pub fn parse(encoded: impl AsRef<[u8]>) -> Result<Self, Base64DecodeError> {
        base64::decode_config(encoded, C::CONF.0).map(Self::new).map_err(Base64DecodeError)
    }
}

impl<C: Base64Config, B: AsRef<[u8]>> fmt::Debug for Base64<C, B> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        self.encode().fmt(f)
    }
}

impl<C: Base64Config, B: AsRef<[u8]>> fmt::Display for Base64<C, B> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        self.encode().fmt(f)
    }
}

impl<'de, C: Base64Config> Deserialize<'de> for Base64<C> {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        let encoded = super::deserialize_cow_str(deserializer)?;
        Self::parse(&*encoded).map_err(de::Error::custom)
    }
}

impl<C: Base64Config, B: AsRef<[u8]>> Serialize for Base64<C, B> {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        serializer.serialize_str(&self.encode())
    }
}

/// An error that occurred while decoding a base64 string.
#[derive(Clone)]
pub struct Base64DecodeError(base64::DecodeError);

impl fmt::Debug for Base64DecodeError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        self.0.fmt(f)
    }
}

impl fmt::Display for Base64DecodeError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        self.0.fmt(f)
    }
}

impl std::error::Error for Base64DecodeError {}
