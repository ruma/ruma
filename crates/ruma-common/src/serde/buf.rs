use bytes::BufMut;
use serde::Serialize;

/// Converts a byte slice to a buffer by copying.
pub fn slice_to_buf<B: Default + BufMut>(s: &[u8]) -> B {
    let mut buf = B::default();
    buf.put_slice(s);
    buf
}

/// Creates a buffer and writes a serializable value to it.
pub fn json_to_buf<B: Default + BufMut, T: Serialize>(val: &T) -> serde_json::Result<B> {
    let mut buf = B::default().writer();
    serde_json::to_writer(&mut buf, val)?;
    Ok(buf.into_inner())
}
