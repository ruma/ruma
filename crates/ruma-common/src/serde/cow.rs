use std::{borrow::Cow, str};

use serde::de::{self, Deserialize, Deserializer, Unexpected, Visitor};

pub(crate) struct MyCowStr<'a>(Cow<'a, str>);

impl<'a> MyCowStr<'a> {
    pub(crate) fn get(self) -> Cow<'a, str> {
        self.0
    }
}

impl<'de> Deserialize<'de> for MyCowStr<'de> {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        deserialize_cow_str(deserializer).map(Self)
    }
}

/// Deserialize a `Cow<'de, str>`.
///
/// Different from serde's implementation of `Deserialize` for `Cow` since it borrows from the
/// input when possible.
///
/// This will become unnecessary if Rust gains lifetime specialization at some point; see
/// <https://github.com/serde-rs/serde/issues/1497#issuecomment-716246686>.
pub fn deserialize_cow_str<'de, D>(deserializer: D) -> Result<Cow<'de, str>, D::Error>
where
    D: Deserializer<'de>,
{
    deserializer.deserialize_string(CowStrVisitor)
}

struct CowStrVisitor;

impl<'de> Visitor<'de> for CowStrVisitor {
    type Value = Cow<'de, str>;

    fn expecting(&self, formatter: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        formatter.write_str("a string")
    }

    fn visit_borrowed_str<E>(self, v: &'de str) -> Result<Self::Value, E>
    where
        E: de::Error,
    {
        Ok(Cow::Borrowed(v))
    }

    fn visit_borrowed_bytes<E>(self, v: &'de [u8]) -> Result<Self::Value, E>
    where
        E: de::Error,
    {
        match str::from_utf8(v) {
            Ok(s) => Ok(Cow::Borrowed(s)),
            Err(_) => Err(de::Error::invalid_value(Unexpected::Bytes(v), &self)),
        }
    }

    fn visit_str<E>(self, v: &str) -> Result<Self::Value, E>
    where
        E: de::Error,
    {
        Ok(Cow::Owned(v.to_owned()))
    }

    fn visit_string<E>(self, v: String) -> Result<Self::Value, E>
    where
        E: de::Error,
    {
        Ok(Cow::Owned(v))
    }

    fn visit_bytes<E>(self, v: &[u8]) -> Result<Self::Value, E>
    where
        E: de::Error,
    {
        match str::from_utf8(v) {
            Ok(s) => Ok(Cow::Owned(s.to_owned())),
            Err(_) => Err(de::Error::invalid_value(Unexpected::Bytes(v), &self)),
        }
    }

    fn visit_byte_buf<E>(self, v: Vec<u8>) -> Result<Self::Value, E>
    where
        E: de::Error,
    {
        match String::from_utf8(v) {
            Ok(s) => Ok(Cow::Owned(s)),
            Err(e) => Err(de::Error::invalid_value(Unexpected::Bytes(&e.into_bytes()), &self)),
        }
    }
}
