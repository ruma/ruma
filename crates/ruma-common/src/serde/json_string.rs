//! De-/serialization functions to and from json strings, allows the type to be used as a query
//! string.

use serde::{
    de::{DeserializeOwned, Deserializer, Error as _},
    ser::{Error as _, Serialize, Serializer},
};

/// Serialize the given value as a JSON string.
pub fn serialize<T, S>(value: T, serializer: S) -> Result<S::Ok, S::Error>
where
    T: Serialize,
    S: Serializer,
{
    let json = serde_json::to_string(&value).map_err(S::Error::custom)?;
    serializer.serialize_str(&json)
}

/// Read a string from the input and deserialize it as a `T`.
pub fn deserialize<'de, T, D>(deserializer: D) -> Result<T, D::Error>
where
    T: DeserializeOwned,
    D: Deserializer<'de>,
{
    let s = super::deserialize_cow_str(deserializer)?;
    serde_json::from_str(&s).map_err(D::Error::custom)
}
