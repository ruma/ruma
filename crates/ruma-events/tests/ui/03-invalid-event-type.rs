use ruma_macros::EventContent;
use serde::{Deserialize, Serialize};

#[derive(Clone, Debug, Deserialize, Serialize, EventContent)]
#[not_ruma_event(type = "m.macro.test", kind = State)]
pub struct MacroTest {
    pub test: String,
}

#[derive(Clone, Debug, Deserialize, Serialize, EventContent)]
#[ruma_event(event = "m.macro.test", kind = State)]
pub struct MoreMacroTest {
    pub test: String,
}

fn main() {}
