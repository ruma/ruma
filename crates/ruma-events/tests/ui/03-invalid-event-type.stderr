error: no event type attribute found, add `#[ruma_event(type = "any.room.event", kind = Kind)]` below the event content derive
 --> $DIR/03-invalid-event-type.rs:4:48
  |
4 | #[derive(Clone, Debug, Deserialize, Serialize, EventContent)]
  |                                                ^^^^^^^^^^^^
  |
  = note: this error originates in a derive macro (in Nightly builds, run with -Z macro-backtrace for more info)

error: expected one of: `type`, `kind`, `skip_redaction`, `custom_redacted`
  --> $DIR/03-invalid-event-type.rs:11:14
   |
11 | #[ruma_event(event = "m.macro.test", kind = State)]
   |              ^^^^^

error: cannot find attribute `not_ruma_event` in this scope
 --> $DIR/03-invalid-event-type.rs:5:3
  |
5 | #[not_ruma_event(type = "m.macro.test", kind = State)]
  |   ^^^^^^^^^^^^^^ help: a derive helper attribute with a similar name exists: `ruma_event`
